### PR TITLE
A circuit for computing c, from section 5, step 5 of "A multi-folding scheme for CCS"

### PR DESCRIPTION
**Pull request by dmpierre**
Gadget to compute $\sum_{j \in [t]} \gamma^{j} \cdot e_1 \cdot \sigma_j + \gamma^{t+1} \cdot e_2 \cdot \sum_{i=1} c_i * \prod_{j \in S_i} \theta_j$. This is the sum computed by the verifier and laid out in section 5, step 5 of "A multi-folding scheme for CCS".

Gadget methods defined [here](https://github.com/privacy-scaling-explorations/folding-schemes/blob/da169ce33e3bde5fb56ba5d1c68f2afffd66c30b/src/folding/hypernova/circuit.rs#L85). Test implemented [here](https://github.com/privacy-scaling-explorations/folding-schemes/blob/da169ce33e3bde5fb56ba5d1c68f2afffd66c30b/src/folding/hypernova/circuit.rs#L278).

### CODE REVIEW

**Review by arnaucube — COMMENTED** [review]
Nice work!

I'm wondering if we could maybe merge the `SumMulsGammaPowsEqSigmaGadget`, `SumCiMulProdThetajGadget` into `ComputeCFromSigmasAndThetasGadget`.

The reasoning is that `SumMulsGammaPowsEqSigmaGadget` and `SumCiMulProdThetajGadget` are not used anywere else than inside `ComputeCFromSigmasAndThetasGadget`, so the methods `SumMulsGammaPowsEqSigmaGadget::sum_muls_gamma_pows_eq_sigma()` and `SumCiMulProdThetajGadget::sum_ci_mul_prod_thetaj()` could be private methods of `ComputeCFromSigmasAndThetasGadget`, since at the end they are internal intermediate steps of the `ComputeCFromSigmasAndThetasGadget::compute_c_from_sigmas_and_thetas()` method.
And `ComputeCFromSigmasAndThetasGadget::compute_c_from_sigmas_and_thetas()` being the only public method of the gadget.
In this way we avoid having extra complexname-gadgets exposed publicly.

How do you see it?

**Comment by dmpierre** [comment]
> Nice work!
> 
> I'm wondering if we could maybe merge the `SumMulsGammaPowsEqSigmaGadget`, `SumCiMulProdThetajGadget` into `ComputeCFromSigmasAndThetasGadget`.
> 
> The reasoning is that `SumMulsGammaPowsEqSigmaGadget` and `SumCiMulProdThetajGadget` are not used anywere else than inside `ComputeCFromSigmasAndThetasGadget`, so the methods `SumMulsGammaPowsEqSigmaGadget::sum_muls_gamma_pows_eq_sigma()` and `SumCiMulProdThetajGadget::sum_ci_mul_prod_thetaj()` could be private methods of `ComputeCFromSigmasAndThetasGadget`, since at the end they are internal intermediate steps of the `ComputeCFromSigmasAndThetasGadget::compute_c_from_sigmas_and_thetas()` method. And `ComputeCFromSigmasAndThetasGadget::compute_c_from_sigmas_and_thetas()` being the only public method of the gadget. In this way we avoid having extra complexname-gadgets exposed publicly.
> 
> How do you see it?

Great idea, makes things much cleaner. Done with [4566385](https://github.com/privacy-scaling-explorations/folding-schemes/pull/61/commits/4566385e43ee2fa1849a24ff9e872f8599e84a5f)!

**Review by arnaucube — APPROVED** [review]
LGTM!

File: src/folding/hypernova/circuit.rs [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    /// It corresponds to the first term of the sum that $\mathcal{V}$ has to compute at section 5, step 5 of "A multi-folding scheme for CCS".
+    ///
+    /// # Arguments
+    /// - `sigmas`: vector of $\sigma_j$ values
+    /// - `eq_eval`: the value of $\tilde{eq}(x_j, x^{\prime})$
+    /// - `gamma`: value $\gamma$
+    /// - `j`: the power at which we start to compute $\gamma^{j}$. This is needed in the context of multifolding.
+    ///
+    /// # Notes
+    /// In the context of multifolding, `j` corresponds to `ccs.t` in `compute_c_from_sigmas_and_thetas`
```
Comment (CPerezz) [inline src/folding/hypernova/circuit.rs:32]: We should translate this to it's R1CS equivalent and document it too.

**Review by CPerezz — APPROVED** [review]
LGTM! Nicely done!

**Pull request merged**